File: Sources/CountriesFlagsObjC/source/Countries/UIImage+Ghana.m
//
//  UIImage+Ghana.m
//  CountriesFlags
//

#import "Countries/UIImage+Ghana.h"

#import "Extensions/UIBezierPath+Star.h"

@implementation UIImage (Ghana)

+ (UIImage *)flagGhanaWithSize:(CGSize)size {
    UIColor* colorRed = [UIColor colorWithRed:207.0f/255.0f
                                        green:9.0f/255.0f
                                         blue:33.0f/255.0f
                                        alpha:1.0f];
    UIColor* colorYellow = [UIColor colorWithRed:252.0f/255.0f
                                           green:210.0f/255.0f
                                            blue:15.0f/255.0f
                                           alpha:1.0f];
    UIColor* colorGreen = [UIColor colorWithRed:0.0f
                                          green:107.0f/255.0f
                                           blue:61.0f/255.0f
                                          alpha:1.0f];
    UIColor* colorBlack = [UIColor blackColor];

    UIGraphicsBeginImageContextWithOptions(size, YES, 0.0f);
    
    [colorRed setFill];
    UIRectFill(CGRectMake(0.0f, 0.0f, size.width, size.height));
    
    [colorYellow setFill];
    UIRectFill(CGRectMake(0.0f, size.height / 3.0f, size.width, size.height / 3.0f));
    
    [colorGreen setFill];
    UIRectFill(CGRectMake(0.0f, size.height / 3.0f * 2.0f, size.width, size.height / 3.0f));
    
    [colorBlack setFill];
    CGFloat radius = (size.height / 12.0f) * 4.422291236 / 2.0f;
    CGFloat xDelta = (size.height / 12.0f) * 0.211145618;
    [[UIBezierPath starAtPoint:CGPointMake(size.width / 2.0f, (size.height + xDelta * 2.0f) / 2.0f)
                        radius:radius
                          type:FlagStarTypeFivePointed] fill];
    
    UIImage* image = UIGraphicsGetImageFromCurrentImageContext();
    UIGraphicsEndImageContext();
    
    return image != nil ? image : [UIImage new];
}

@end

File: Sources/CountriesFlags/Countries/UIImage+Ghana.swift
//
//  UIImage+Ghana.swift
//  CountriesFlags
//

#if canImport(UIKit)

import UIKit

extension UIImage {
    /**
     [wikipedia]: https://en.wikipedia.org/wiki/Flag_of_Ghana "wikipedia"
     
     Generates the flag of Ghana.
     For more information see [wikipedia].

     - Parameters:
        - size: image size.
     
     - Returns: Returns the flag of Ghana.
     */
    internal static func flagGhana(with size: CGSize) -> UIImage? {
        let colorRed = UIColor(red: 207.0 / 255.0,
                               green: 9.0 / 255.0,
                               blue: 33.0 / 255.0,
                               alpha: 1.0)
        let colorYellow = UIColor(red: 252.0 / 255.0,
                                  green: 210.0 / 255.0,
                                  blue: 15.0 / 255.0,
                                  alpha: 1.0)
        let colorGreen = UIColor(red: 0.0,
                                 green: 107.0 / 255.0,
                                 blue: 61.0 / 255.0,
                                 alpha: 1.0)
        let colorBlack = UIColor.black

        UIGraphicsBeginImageContextWithOptions(size, false, 0.0)

        colorRed.setFill()
        UIRectFill(CGRect(origin: .zero, size: size))

        colorYellow.setFill()
        UIRectFill(CGRect(origin: CGPoint(x: 0.0, y: size.height / 3.0),
                          size: CGSize(width: size.width, height: size.height / 3.0)))

        colorGreen.setFill()
        UIRectFill(CGRect(origin: CGPoint(x: 0.0, y: size.height / 3.0 * 2.0),
                          size: CGSize(width: size.width, height: size.height / 3.0)))

        colorBlack.setFill()
        let radius = CGFloat(size.height / 12.0) * 4.422291236 / 2.0
        let xDelta = CGFloat(size.height / 12.0) * 0.211145618
        UIBezierPath.star(with: CGSize(width: size.width, height: size.height + xDelta * 2.0),
                          radius: radius).fill()

        let image = UIGraphicsGetImageFromCurrentImageContext()
        UIGraphicsEndImageContext()

        return image
    }
}

#endif
